src/libDirectoryService/ViewChangePreProcessing.cpp

```diff
@@ -294,17 +367,90 @@ bool DirectoryService::ComputeNewCandidateLeader() {
-            m_viewChangeCounter, get_time_as_int(), committeeHash),
+            newLeaderNetworkInfo,
+            m_mediator.m_DSCommittee->at(candidateLeaderIndex).first,
+            m_viewChangeCounter, m_cumulativeFaultyLeaders, get_time_as_int(),
+            committeeHash),
         CoSignatures()));
     m_pendingVCBlock->SetBlockHash(m_pendingVCBlock->GetHeader().GetMyHash());
   }
 
   return true;
 }
 
-bool DirectoryService::RunConsensusOnViewChangeWhenCandidateLeader() {
+uint32_t DirectoryService::CalculateNewLeaderIndex() {
+  // New leader is computed using the following
+  // new candidate leader index is
+  // H((finalblock or vc block), vc counter) % size
+  // of ds committee
+  SHA2<HASH_TYPE::HASH_VARIANT_256> sha2;
+
+  uint64_t latestIndex = m_mediator.m_blocklinkchain.GetLatestIndex();
+  BlockLink bl = m_mediator.m_blocklinkchain.GetBlockLink(latestIndex);
+  VCBlockSharedPtr prevVCBlockptr;
+  if (CheckUseVCBlockInsteadOfDSBlock(bl, prevVCBlockptr)) {
+    LOG_GENERAL(INFO,
+                "Using hash of last vc block for computing candidcate leader");
+    sha2.Update(prevVCBlockptr->GetBlockHash().asBytes());
+  } else {
+    LOG_GENERAL(
+        INFO, "Using hash of last final block for computing candidcate leader");
+    sha2.Update(
+        m_mediator.m_txBlockChain.GetLastBlock().GetBlockHash().asBytes());
+  }
+
+  vector<unsigned char> vcCounterBytes;
+  Serializable::SetNumber<uint32_t>(vcCounterBytes, 0, m_viewChangeCounter,
+                                    sizeof(uint32_t));
+  sha2.Update(vcCounterBytes);
+  uint16_t lastBlockHash = DataConversion::charArrTo16Bits(sha2.Finalize());
+  uint32_t candidateLeaderIndex =
+      lastBlockHash % (m_mediator.m_DSCommittee->size());
+
+  while (candidateLeaderIndex == m_consensusLeaderID) {
+    LOG_GENERAL(INFO,
+                "Computed candidate leader is current faulty ds leader. Index: "
+                    << candidateLeaderIndex);
+    sha2.Update(sha2.Finalize());
+    lastBlockHash = DataConversion::charArrTo16Bits(sha2.Finalize());
+    candidateLeaderIndex = lastBlockHash % (m_mediator.m_DSCommittee->size());
+    LOG_GENERAL(INFO, "Re-computed candidate leader is at index: "
+                          << candidateLeaderIndex
+                          << " VC counter: " << m_viewChangeCounter);
+  }
+  sha2.Reset();
+  return candidateLeaderIndex;
+}
+
+bool DirectoryService::CheckUseVCBlockInsteadOfDSBlock(
+    const BlockLink& bl, VCBlockSharedPtr& prevVCBlockptr) {
+  BlockType latestBlockType = get<BlockLinkIndex::BLOCKTYPE>(bl);
+
+  if (latestBlockType == BlockType::VC) {
+    if (!BlockStorage::GetBlockStorage().GetVCBlock(
+            get<BlockLinkIndex::BLOCKHASH>(bl), prevVCBlockptr)) {
+      LOG_GENERAL(WARNING, "could not get vc block "
+                               << get<BlockLinkIndex::BLOCKHASH>(bl));
+      return false;
+    }
+
+    if (prevVCBlockptr->GetHeader().GetViewChangeEpochNo() !=
```

Comment:
Should it be `GetViewChangeDSEpochNo()`?

Reply:
Fixed